django/db/models/fields/generated.py

```diff
@@ -68,11 +68,55 @@ def generated_sql(self, connection):
 
     def check(self, **kwargs):
         databases = kwargs.get("databases") or []
-        return [
+        errors = [
             *super().check(**kwargs),
             *self._check_supported(databases),
             *self._check_persistence(databases),
         ]
+        output_field_clone = self.output_field.clone()
+        output_field_clone.model = self.model
+        output_field_checks = output_field_clone.check(databases=databases)
+        if output_field_checks:
+            if not self._output_field:
+                errors.append(
+                    checks.Error(
+                        "You must set GeneratedField.output_field as it cannot be "
+                        "inferred.",
+                        obj=self,
+                        id="fields.E222",
+                        hint="Set output_field on the field.",
+                    )
+                )
```

Comment:
Simon's [proposition](https://github.com/django/django/pull/17464#issuecomment-1807216304) is rather to make `GeneratedField.output_field` mandatory (will work on this tomorrow.)

Reply:
A check work as well but I think it's safer to _encode_ in the function signature for the time being.

Forcing it to be specified ensures that if we make improvements to the ORM `output_field` resolving we don't end up inadvertently changing the SQL generated for the column. This is something that has bit us already a few times with expressions and their generated SQL changing between versions and making indices unusable.

Having the `output_field` explicit in historical migrations ensures we can more freely make adjustments on this front.

Reply:
Agreed.